### PR TITLE
Course Admin Modal

### LINKED ISSUE #229: Course admin interface: create new / edit modal
See modal mockup in https://seas-computing.github.io/course-planner_ux/manage_courses.html
Utilises modal component from MarkOne. https://seas-computing.github.io/mark-one/#/Modals

* [x] Create Course modal opens a blank form
  * "Same as" is just plain text, without the + sign
  * Undergraduate is false by default (ask Kathy and Patrick about whether it should be true)
  * "Areas" drop-down populated with list of all the areas
  * Look at stacking all fields
  * Change buttons from "Create" to "Save"
  * Clicking "Save" submits the form via existing POST method (#135 )
    * Table updates with new faculty on successful submit
* [x] Edit faculty modal opens a populated form
  * Table updates with new faculty details on successful submit
  * Clicking "Save" submits the form via existing PUT method (#136)
* [x] Implement Dropdown refactor changes on both this and faculty admin modals.
* [x] Docs
* [x] Tests
* [x] [Accessibility](https://github.com/seas-computing/course-planner/issues/214#issuecomment-627472321)

### PR DESCRIPTION
**Pull request by natalynyu**
This PR adds the course admin modal, which allows users to create and edit courses. The `GET` of the course admin table was updated so that the entries would be sorted appropriately. The client side API was also added in this PR to create and edit courses. The `Course` entity was also updated to add in the `cascade: insert` option for areas so that when a new area is created by the user, it is created in the database.

I would recommend making sure that you are able to create and edit a course and that the new area appears in the dropdown after having created a new area. We decided as a group in the sprint planning meeting to add the new area in the `metadata.areas` array so that the new area appears while the user is in the current session. The alternative was to make another `GET` request to get all of the areas in the database, but we decided given the small number of admins who would be using the app at once.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality
      to change)

## Checklist:

- [x] I have run `eslint` on the code
- [x] I have added JSDoc for all of my code (where applicable)
- [x] I have added tests to cover my changes.

## Priority:
- [x] Normal 
- [ ] High 

## Related Issues:

Fixes #229

### CODE REVIEW

File: src/common/dto/courses/CreateCourse.dto.ts [hunk 1]
```diff
@@ -33,56 +32,51 @@ export abstract class CreateCourse {
...
-  @IsOptional()
-  public notes?: string;
-
   @ApiModelProperty({
     type: 'boolean',
     example: false,
   })
-  @IsNotEmpty()
   @IsBoolean()
+  @IsOptional()
```
Comment (rmainseas) [inline src/common/dto/courses/CreateCourse.dto.ts:79]: If this is optional - what's the default value of it?

**Review by rmainseas — CHANGES_REQUESTED** [review]

File: src/common/dto/util.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// A helper function that removes the whitespace from the beginning and end of
+// a string. This will be used to sanitize text input submitted by the user.
+export const trimIfString = (value: string): string => value?.trim();
```
[the remark this answers is not in the document]
Comment (natalynyu) [reply src/common/dto/util.ts]: image

Thanks for pointing this out! I ended up changing it to `(value = '')` due to the error message above.

File: src/common/dto/__tests__/util.test.tsx [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+      it('should return the original string', function () {
+        const stringWithoutWhitespace = physicsCourseResponse.catalogNumber;
+        strictEqual(
+          trimIfString(stringWithoutWhitespace),
+          physicsCourseResponse.catalogNumber
+        );
+      });
+    });
+    context('when no value is passed to the function', function () {
+      it('should return undefined', function () {
```
[the remark this answers is not in the document]
Comment (natalynyu) [reply src/common/dto/__tests__/util.test.tsx]: Good point. This was also updated in 6b5896e.

File: src/common/utils/util.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+// A helper function to convert a string from camel case to title case
```
[the remark this answers is not in the document]
Comment (natalynyu) [reply src/common/utils/util.ts]: Good idea. I updated this comment in commit 90630b7.

File: src/common/dto/util.ts [hunk 5]
```diff
@@ -0,0 +1,3 @@
+// A helper function that removes the whitespace from the beginning and end of
+// a string. This will be used to sanitize text input submitted by the user.
```
[the remark this answers is not in the document]
Comment (natalynyu) [reply src/common/dto/util.ts]: This was also updated in 90630b7.

File: src/common/dto/courses/CreateCourse.dto.ts (hunk 1 above)
Comment (natalynyu) [reply src/common/dto/courses/CreateCourse.dto.ts:79]: I thought that the default was being set to `true` by this code in the course entity? Should I not have the `@IsOptional` decorator?

Comment (rmainseas) [reply src/common/dto/courses/CreateCourse.dto.ts:79]: Sure - I was more meaning that you might want to add the default value in to the `@ApiModelProperty` decorator for the documentation :)

Comment (natalynyu) [reply src/common/dto/courses/CreateCourse.dto.ts:80]: Ah, I see, ok, thanks! I updated the `@ApiModelProperty` for the create and update course DTOs in 905b8f8.

**Review by rmainseas — APPROVED** [review]
This is good to go.

**Pull request merged**